### PR TITLE
ARROW-4544: [Rust] JSON nested struct reader

### PR DESCRIPTION
**Pull request by nevi-me**
This allows reading nested JSON data provided in a schema. If the expected value doesn't exist, or is not an object, it is deemed to be null.

I tested that the implementation correctly creates a null bitmap for the struct.

This does not include schema inference.

### CODE REVIEW

**Comment by nevi-me** [comment]
CC @houqp

File: rust/arrow/src/json/reader.rs [hunk 1]
```diff
@@ -1022,6 +846,228 @@ impl<R: Read> Reader<R> {
...
+                            }
+                        }
+                        DataType::Dictionary(ref key_ty, ref val_ty) => self
+                            .build_string_dictionary_array(
+                                rows,
+                                field.name(),
+                                key_ty,
+                                val_ty,
+                            ),
+                        DataType::Struct(fields) => {
```
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:1022]: This is the main change, we recurse into the `build_struct_array` function

File: rust/arrow/src/json/reader.rs [hunk 2]
```diff
@@ -1022,6 +846,228 @@ impl<R: Read> Reader<R> {
...
+                                        i,
+                                        row.as_object()
+                                            .map(|v| v.get(field.name()))
+                                            .flatten(),
+                                    )
+                                })
+                                .map(|(i, v)| match v {
+                                    // we want the field as an object, if it's not, we treat as null
+                                    Some(Value::Object(value)) => {
+                                        bit_util::set_bit(null_buffer.data_mut(), i);
```
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:1042]: It should be more optimal to set all values as true, then only set the non-objects to false. I was struggling to get this right though, don't know why the test was failing.

**Comment by andygrove** [comment]
I haven't reviewed in detail but the changes look logical and make sense to me.

File: rust/arrow/src/json/reader.rs [hunk 3]
```diff
@@ -1559,6 +1608,54 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_nested_struct_json_arrays() {
```
Comment (alamb) [inline rust/arrow/src/json/reader.rs:1612]: This is a pretty cool test case demonstrating what is going on -- TIL I learn that we have a JSON schema extractor in arrow!

File: rust/arrow/test/data/nested_structs.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{"a": {"b": true, "c": {"d": "text"}}}
```
Comment (alamb) [inline rust/arrow/test/data/nested_structs.json:1]: Given the size of this file (very small) I wonder if the test case itself would be more readable if the contents were inlined directly into the rust source code. I don't see any issues with it being in its own file, this is more a style comment

**Review by alamb — APPROVED** [review]

**Review by paddyhoran — APPROVED** [review]

File: rust/arrow/test/data/nested_structs.json (hunk 4 above)
Comment (nevi-me) [reply rust/arrow/test/data/nested_structs.json:1]: I opened https://issues.apache.org/jira/browse/ARROW-10764

**Pull request closed without merging**